### ISSUE TITLE
False positives when using `transmute()` with raw trait object pointers to change trait object lifetimes

### ISSUE THREAD

**Issue by okready** [rust-lang/rust-clippy#2906]
With clippy 0.0.212, I'm getting what appears to be false positives when using `transmute()` to cast either to or from a raw trait object pointer where the input and output types have invariant lifetimes, or if the output type has a longer lifetime such as `'static` (casting using `as` isn't allowed in such cases). No warnings are triggered if the source and target types are both references.

```rust
use std::mem::transmute;

pub trait DummyTrait {}

pub unsafe fn ref_to_ref<'a, 'b>(obj: &'a (DummyTrait + 'a)) -> &'b (DummyTrait + 'b) {
    // No warnings.
    transmute(obj)
}

pub unsafe fn ptr_to_ptr<'a, 'b>(obj: *const (DummyTrait + 'a)) -> *const (DummyTrait + 'b) {
    // Triggers `transmute_ptr_to_ptr` warning.
    transmute(obj)
}

pub unsafe fn ptr_to_ref<'a, 'b>(obj: *const (DummyTrait + 'a)) -> &'b (DummyTrait + 'b) {
    // Triggers `transmute_ptr_to_ref` warning.
    transmute(obj)
}

pub unsafe fn ref_to_ptr<'a, 'b>(obj: &'a (DummyTrait + 'a)) -> *const (DummyTrait + 'b) {
    // Triggers `useless_transmute` warning.
    transmute(obj)
}
```

**Comment by rail-rain** [comment]
The warning has changed due to #5343. Here is the lastest one for your convinience.

```rust
#![allow(clippy::missing_safety_doc)]
// `useless_transmute` is now an allowed-by-default lint.
#![warn(clippy::useless_transmute)]

use std::mem::transmute;

pub trait DummyTrait {}

pub unsafe fn ref_to_ref<'a, 'b>(obj: &'a (dyn DummyTrait + 'a)) -> &'b (dyn DummyTrait + 'b) {
    // Triggers `useless_transmute` warning.
    // Note: this is the false positive introduced #5343
    transmute(obj)
}

pub unsafe fn ptr_to_ptr<'a, 'b>(obj: *const (dyn DummyTrait + 'a)) -> *const (dyn DummyTrait + 'b) {
    // Triggers `useless_transmute` warning.
    // Note: `useless_transmute` took over because `useless_transmute` has a higher priority than `transmute_ptr_to_ptr`
    // besides the false positive above.
    transmute(obj)
}

pub unsafe fn ptr_to_ref<'a, 'b>(obj: *const (dyn DummyTrait + 'a)) -> &'b (dyn DummyTrait + 'b) {
    // Triggers `transmute_ptr_to_ref` warning.
    transmute(obj)
}

pub unsafe fn ref_to_ptr<'a, 'b>(obj: &'a (dyn DummyTrait + 'a)) -> *const (dyn DummyTrait + 'b) {
    // Triggers `useless_transmute` warning.
    transmute(obj)
}
```

This issue means (every?) lints for `transmute` don't account lifetimes, including `transmutes_expressible_as_ptr_cast` (if you commented out other `transmute` related lints, it would fire on the example).
